src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,51 +31,102 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) {
+    std::shared_ptr<ValueSegment<T>> value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(base_segment);
+
+    // create dictionary
+    _dictionary = std::make_shared<std::vector<T>>(value_segment->values());
+    std::sort(_dictionary->begin(), _dictionary->end());
+    // eliminates all but the first element from every consecutive group of equivalent elements
+    auto last = std::unique(_dictionary->begin(), _dictionary->end());
+    // the dictionary must be resized following the unique operation
+    _dictionary->erase(last, _dictionary->end());
+
+    // create Attribute Vector with the most fitting width
+    if (_dictionary->size() <= std::numeric_limits<uint8_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(value_segment->size());
+    } else if (_dictionary->size() <= std::numeric_limits<uint16_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(value_segment->size());
+    } else {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(value_segment->size());
+    }
+
+    // for each entry in the segment
+    for (ColumnID segment_column(0); segment_column < value_segment->size(); ++segment_column) {
+      // find corresponding dictionary value
+      const ValueID& dictionary_value_id = lower_bound(type_cast<T>((*value_segment)[segment_column]));
+      DebugAssert(dictionary_value_id != INVALID_VALUE_ID, "a dictionary value must be found for each attribute");
+      _attribute_vector->set(segment_column, dictionary_value_id);
+    }
+  }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    PerformanceWarning("operator[] used");
+
+    auto valueID = _attribute_vector->get(chunk_offset);
+    return _dictionary->at(valueID);
+  }
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const {
+    auto valueID = _attribute_vector->get(chunk_offset);
+    return _dictionary.at(valueID);
```

Comment:
This should be `return _dictionary->at(valueID);` since you are accessing a pointer right? Even if the code works, you should probabily change it for consistency with the rest of the code

Reply:
Thanks a lot for this hint. Fixed.